capsules/src/alarm.rs

```diff
@@ -248,14 +248,14 @@ impl<'a, A: Alarm<'a>> SyscallDriver for AlarmDriver<'a, A> {
 
 impl<'a, A: Alarm<'a>> time::AlarmClient for AlarmDriver<'a, A> {
     fn alarm(&self) {
-        let now: Ticks32 = Ticks32::from(self.alarm.now().into_u32());
+        let now = self.alarm.now();
```

Comment:
I am a little worried about this change. `Expiration` stores values in 32 bits.  But `now` can return a 64-bit value. So what happens on line 256 below if now > 2^32? I think the calculation will be incorrect. For example, suppose

```
reference       = 0xffffff00
dt              = 0x00000110
reference + dt  = 0x00000010
now            = 0x100000020
```

Technically, `now` is after `reference + dt`, so it's not within their range. However, since reference +  dt is considered in the 64-bit space, it will be. This change of moving to the `Ticks` type only works if `Expiration` is also in terms of the `Ticks` type, which is why I chose to stick with 32 bits.

Reply:
Good catch!

My original PR did update `Expiration` so I missed this. It still would have broken bisectability though, so it needs to be fixed.

Wouldn't something like this fix the issue? That way `now` and `reference + dt` will be the same:

```diff
diff --git a/capsules/src/alarm.rs b/capsules/src/alarm.rs
index 6900c1da1..82a808840 100644
--- a/capsules/src/alarm.rs
+++ b/capsules/src/alarm.rs
@@ -255,7 +255,7 @@ impl<'a, A: Alarm<'a>> time::AlarmClient for AlarmDriver<'a, A> {
                 // as passed (since reference must be in the past)
                 if !now.within_range(
                     A::Ticks::from(reference),
-                    A::Ticks::from(reference.wrapping_add(dt)),
+                    A::Ticks::from((reference as u64).wrapping_add(dt as u64)),
                 ) {
                     alarm.expiration = Expiration::Disabled;
                     self.num_armed.set(self.num_armed.get() - 1);
```

It also aligns with the future `u64` value for `Expiration`

Reply:
This clearly doesn't work. What happens when `now` is significantly greater than 2^32, i.e. well after the first 32-bit increment? Since `reference` <= 2^32, I this means `now` will not fall within `reference + dt`.

There shouldn't be a plan to change `Expiration` to `u64`. The [TRD105](https://github.com/tock/tock/blob/master/doc/reference/trd105-time.md) is clear that `Alarm` is provided for 32 bits but `Time` is provided for 64 bits. If we want a 64-bit alarm interface that should become a new device/system call driver.

Reply:
I think at this point it might be good for you to set up a testing configuration. When I redesigned the time system, I'd test by running timer tests for a few days and checking the output. I'd also test 32-bit wraparound more heavily by setting the counter register value to close to wraparound.

Reply:
Hmm... If we aren't allowed a `u64` in `Expiration` it starts to become really hard. As our `reference` will always possibly be less then the current time.

I have pushed an update that handles a `now` greater then 2^32. I'll run some tests on it as well

Reply:
Tock the system call driver provides 32-bit alarms, but it can support 64-bit time. Why change it to propagate the underlying alarm width into its logic?

Reply:
The current time can be a 64-bit value, so we want to use that for setting alarms, otherwise we might alarm at the wrong time if we ignore the high bits of the current time

Reply:
You don't ignore the high bits. The TRD is pretty clear that it's 32-bit alarms and 64-bit time.

Reply:
The current implementation is a 32-bit time. This line:

```rust
let now: Ticks32 = Ticks32::from(self.alarm.now().into_u32());
```

gets the current time as a 32-bit value. So we need a change like this to instead use a 64-bit time. The same applies for `reference` which is `now` when we schedule the alarm (so a 64-bit time)